#167733644 signup/Registration Validation Errors

dchima opened the pull request:
### What does this PR do?
Validate user inputs and throws errors whenever user does not meet input requirements.

### Description of Task to be completed?
- Install needed validation depenencies (e.g. @hapi/joi package)
- Create user validation middleware
-  Add validation class for user input at signup
- Update routes
- Write the test for validations
- Test validations (on success and on failure)

### How should this be manually tested?
- Clone repository
- update or copy .env file from .env.sample
- Run `npm run prestart` to create tables in your database.
- Run  npm run start:dev on local root directory
- Test router on Postman with POST: 'localhost:3000/api/auth/signup'
- Run a test case JSON script on Postman body, eg:
```
{
	"firstName": "Tony",
	"lastName": "Marshal",
	"email": "tony@marshal.com",
	"password": "tmonarqA1.",
	"gender": "male",
	"street": "Abage",
	"city": "Lekki",
	"state": "Lagos",
	"country": "Nigeria",
	"birthdate": "2009-08-20",
	"phoneNumber": "(+234) 098-3442-1889",
	"companyName": "Jubilee"
}
```
- Omit some fields to check that the validation script works on inputs not given by the user.

### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#167733644](https://www.pivotaltracker.com/story/show/167733644)

dchima committed:
ft(joi): added @hapi/joi package

[Starts #167733644]

dchima committed:
refactor(response): added a function for handling response objects

[onging #167733644]

dchima committed:
ft(validation) added validation class for user inputs

This commit adds a validation class for user registration inputs.
upon failure, the class returns the relevant validation errors
back to the client

[Onging #167733644]

dchima committed:
ft(validation): adjusted minimum password length

[Onging #167733644]

dchima committed:
chore(test): added testing for user input validation checks

[Ongoing #167733644]

dchima committed:
fix(conflict): finishing fixes to avoid merge conflicts

[Ongoing #167733644]

dchima committed:
bg(test): fixed testing bug on phone number

Changed phone number value on testing and added loggers
Also added comments

[Ongoing #167733644]

dchima committed:
ft(validation): added database check and number validation

This commit adds validations to check if the user
exists in the database upon signup.
It also adds phone number valdations for international
number formats

[Onging #167733644]

dchima committed:
chore(test) added extra test for exixsting user signup

[Ongoing #167733644]

dchima committed:
fix(validation): adjusted constraints on validations

constraints adjusted for country and city strings

[Ongoing #167733644]

dchima commented:
Forgot to remove the loggers I used for testing. I'll do that in the next commit

dchima committed:
chore(log): removed console logs

[Ongoing #167733644]

dchima committed:
chore(test): removed unised variable in test file

[Ongoing #167733644]

Segun-Ogundipe commented on src/routes/api/users.js:
```diff
@@ -1,8 +1,9 @@
 const userRouter = express.Router();
 
-userRouter.post('/auth/signup', UserController.createUser);
+userRouter.post('/auth/signup', userValidation.signup, UserController.createUser);
```
This should be in the middleware folder

Segun-Ogundipe replied:
Should be in middleware folder

Segun-Ogundipe commented on src/utils/response.js:
```diff
@@ -0,0 +1,18 @@
+// Response function for error / success returns
```
A response object was previously added to the project.

dchima replied:
I'll remove the redundant response object

Segun-Ogundipe commented on src/validation/userValidation.js:
```diff
@@ -0,0 +1,96 @@
+     * @param {object} next - the returned values going into the next operation.
+     * @returns {object} - returns an object (error or response).
+     */
+  static async signup(req, res, next) {
```
Validations in the validation folder should not be used as a middleware. Validation methods should be created here and imported in a middleware file.

King-Success commented on src/validation/userValidation.js:
```diff
@@ -0,0 +1,96 @@
+import UserService from '../services/index';
+
+// password complexity object
+const complexityOPtions = {
```
notice typo in complexityOptions, P should be lower case

Segun-Ogundipe reviewed (changes requested):
Looks good.
There's no Travis build presently running. Travis build should be fixed before merging

dchima committed:
refactor(middleware): added validation to middleware file

Added a middleware class to call the validations it needs
also removed redundant methods (response.js)

[Ongoing #167733644]

dchima committed:
chore(response) removed response call

[Ongoing #167733644]

Segun-Ogundipe commented on src/middlewares/index.js:
```diff
@@ -1,3 +1,4 @@
+import userMiddleware from './userMiddleware';
 
-export default TokenMiddleware;
+export { TokenMiddleware, userMiddleware };
```
The TTL said we should prefer default export

dchima replied:
Alright, I'll make the changes.

Segun-Ogundipe commented on src/middlewares/userMiddleware.js:
```diff
@@ -0,0 +1,17 @@
+     * @returns {object} - returns an object (error or response).
+     */
+  static onSignup(req, res, next) {
+    userValidation.signup(req, res, next);
```
Uservalidation should either throw an error in the form of ApiError when there's an error in validation or return true when all validation passes.
Sending back responses to the user should be done in the middleware.
The method in the middleware could be renamed to be more descriptive e.g `static validateSignup()`

dchima replied:
The error responses are already handled in the userValidation call. I don't see why this has to be done specifically in the middleware class. All error handling occurs in the validation class already (on success and on failure).

Segun-Ogundipe commented on src/validation/userValidation.js:
```diff
@@ -39,7 +39,7 @@ export default class userValidation {
-      password: new passwordComplexity(complexityOPtions).required()
+      password: new passwordComplexity(complexityOptions).required()
         .label('invalid password format'),
       gender: joi.string().valid('male', 'female').required()
```
The error messages are not descriptive.
For instance, the same error is being sent to the user if the password was not provided or if the format is invalid
The error should be specific. `password is required` when the password is missing `password should include...` when the format is wrong

dchima replied:
I'll do just that. However, this doesn't need much recoding, just a change on how descriptive the password error is. If a password isn't inputted, a label like "Password is required. It should be more than 8 characters, and should include at least a capital letter, and a number" should cover both cases in a single error message.

Segun-Ogundipe reviewed (changes requested):
Good job with separating the concerns!!!
I left a few comments on some files.
In the validation, an error should be thrown when validation fails with a descriptive error message. This could be achieved with `new ApiError(400, error.details)` and `return true` when validation passes, then in the middleware, you could:
```try{
       const isValid = userValidation(req.body);

       if(isValid === true) next;
   } catch (error) {
     res.status(error.status || 500).json(new ApiResponse(error.status || 500, error.message));
   }
```
This way, you won't have to send responses in the validator

dchima commented:
> Good job with separating the concerns!!!
> I left a few comments on some files.
> In the validation, an error should be thrown when validation fails with a descriptive error message. This could be achieved with `new ApiError(400, error.details)` and `return true` when validation passes, then in the middleware, you could:
> 
> ```
>        const isValid = userValidation(req.body);
> 
>        if(isValid === true) next;
>    } catch (error) {
>      res.status(error.status || 500).json(new ApiResponse(error.status || 500, error.message));
>    }
> ```
> 
> This way, you won't have to send responses in the validator

Alright, I'll incorporate this in the next commit tonight.

dchima committed:
refactor(validation) moved api response validation class

THe validation class was moved to the middleware call class

[Ongoing #167733644]

dchima committed:
chore(conflicts): resolved merge conflicts with develop branch

[Finishes #167733644]

dchima committed:
chore(eslint): removed unused object

[FInishes #167733644]
